File: backend/backend/legacy/cs_chess_algorithms (pre-python-refactorization)/move_validation_data_algorithms.cs
public void Retrieve_FEN_Notation_Information(ref MoveInfo moveInfo, string fenNotationMove)
{
    // Split the FEN notation into its 6 parts
    string[] fenParts = fenNotationMove.Split(' ');

    // Parse board state
    string boardState = fenParts[0];
    string[] rows = boardState.Split('/');
    for (int y = 0; y < 8; y++)
    {
        int x = 0;
        foreach (char symbol in rows[y])
        {
            if (char.IsDigit(symbol))
            {
                x += (int)char.GetNumericValue(symbol); // Empty squares
            }
            else
            {
                bool isWhite = char.IsUpper(symbol);
                ChessPiece pieceType;
                switch (char.ToLower(symbol))
                {
                    case 'p': pieceType = ChessPiece.Pawn; break;
                    case 'r': pieceType = ChessPiece.Rook; break;
                    case 'n': pieceType = ChessPiece.Knight; break;
                    case 'b': pieceType = ChessPiece.Bishop; break;
                    case 'q': pieceType = ChessPiece.Queen; break;
                    case 'k': pieceType = ChessPiece.King; break;
                    default: throw new Exception("Invalid piece type in FEN.");
                }
                moveInfo.chessBoard[x, 7 - y] = new ChessFigure(pieceType, isWhite);
                x++;
            }
        }
    }

    // Parse active color
    moveInfo.whiteTurn = (fenParts[1] == "w");

    // Parse castling rights
    moveInfo.canWhiteSCastle = fenParts[2].Contains("K");
    moveInfo.canWhiteLCastle = fenParts[2].Contains("Q");
    moveInfo.canBlackSCastle = fenParts[2].Contains("k");
    moveInfo.canBlackLCastle = fenParts[2].Contains("q");

    // Parse en passant target square
    if (fenParts[3] == "-")
    {
        moveInfo.enPassantTarget = (null, null);
    }
    else
    {
        int enPassantX = fenParts[3][0] - 'a';
        int enPassantY = fenParts[3][1] - '1';
        moveInfo.enPassantTarget = (enPassantX, enPassantY);
    }

    // Parse half-move clock and full-move number
    moveInfo.halfMoveClock = int.Parse(fenParts[4]);
    moveInfo.moveNumber = int.Parse(fenParts[5]);
}


public void Retrieve_Coordinate_Notation_Information(ref MoveInfo moveInfo, string coordinateNotationMove)
{
    // Extract source and target positions from the notation (e.g., "b5-c6" or "e7-e8-Q")
    string[] moveParts = coordinateNotationMove.Split('-');
    
    // Parse source position
    moveInfo.sourceX = moveParts[0][0] - 'a'; // 'a' is 0, 'b' is 1, etc.
    moveInfo.sourceY = moveParts[0][1] - '1'; // '1' is 0, '2' is 1, etc.

    // Parse target position
    moveInfo.targetX = moveParts[1][0] - 'a';
    moveInfo.targetY = moveParts[1][1] - '1';

    // If the move contains a promotion (e.g., "e7-e8-Q"), parse the promotion source
    if (moveParts[1].Length == 3)
    {
        char promotionChar = moveParts[1][2];
        switch (promotionChar)
        {
            case 'Q': moveInfo.promotedTo = ChessPiece.Queen; break;
            case 'R': moveInfo.promotedTo = ChessPiece.Rook; break;
            case 'B': moveInfo.promotedTo = ChessPiece.Bishop; break;
            case 'N': moveInfo.promotedTo = ChessPiece.Knight; break;
            default: throw new Exception("Invalid promotion sourcePiece in coordinate notation.");
        }
    }

    // Determine the piece type from the source position and FEN board
    moveInfo.pieceType = moveInfo.chessBoard[moveInfo.sourceX, moveInfo.sourceY].type;
}

public void Retrieve_Aditional_Move_Information(ref MoveInfo moveInfo)
{
    moveInfo.sourcePiece = moveInfo.chessBoard[moveInfo.sourceX, moveInfo.sourceY];
    moveInfo.targetPiece = moveInfo.chessBoard[moveInfo.targetX, moveInfo.targetY];
    moveInfo.distanceX = moveInfo.targetX - moveInfo.targetX;
    moveInfo.distanceY = moveInfo.targetY - moveInfo.targetY;
    moveInfo.directionX = Math.Sign(moveInfo.distanceX);
    moveInfo.directionY = Math.Sign(moveInfo.distanceY);
}

File: backend/backend/python_chess_algorithms/move_validation_data_algorithms.py
from .data_structures import ChessPiece, ChessFigure, MoveInfo
import math

def retrieve_fen_notation_information(move_info, fen_notation_move):
    fen_parts = fen_notation_move.split(' ')

    board_state = fen_parts[0]
    rows = board_state.split('/')
    for y in range(8):
        x = 0
        for symbol in rows[y]:
            if symbol.isdigit():
                x += int(symbol)
            else:
                is_white = symbol.isupper()
                symbol = symbol.lower()
                if symbol == 'p':
                    piece_type = ChessPiece.Pawn
                elif symbol == 'r':
                    piece_type = ChessPiece.Rook
                elif symbol == 'n':
                    piece_type = ChessPiece.Knight
                elif symbol == 'b':
                    piece_type = ChessPiece.Bishop
                elif symbol == 'q':
                    piece_type = ChessPiece.Queen
                elif symbol == 'k':
                    piece_type = ChessPiece.King
                else:
                    raise ValueError("Invalid piece type in FEN.")
                
                move_info.chess_board[x][y] = ChessFigure(piece_type, is_white)
                x += 1

    move_info.white_turn = (fen_parts[1] == 'w')

    move_info.can_white_s_castle = 'K' in fen_parts[2]
    move_info.can_white_l_castle = 'Q' in fen_parts[2]
    move_info.can_black_s_castle = 'k' in fen_parts[2]
    move_info.can_black_l_castle = 'q' in fen_parts[2]

    if fen_parts[3] == '-':
        move_info.en_passant_target = (None, None)
    else:
        en_passant_x = ord(fen_parts[3][0]) - ord('a')
        en_passant_y = 8 - int(fen_parts[3][1])
        move_info.en_passant_target = (en_passant_x, en_passant_y)

    move_info.half_move_clock = int(fen_parts[4])
    move_info.move_number = int(fen_parts[5])


def retrieve_coordinate_notation_information(move_info, coordinate_notation_move):
    move_info.source_x = int(coordinate_notation_move[0])
    move_info.source_y = int(coordinate_notation_move[1])

    move_info.target_x = int(coordinate_notation_move[2])
    move_info.target_y = int(coordinate_notation_move[3])

    move_info.piece_type = move_info.chess_board[move_info.source_x][move_info.source_y].type

    if len(coordinate_notation_move) == 5:
        promotion_char = coordinate_notation_move[4]
        if promotion_char == 'q':
            move_info.promoted_to = ChessPiece.Queen
        elif promotion_char == 'r':
            move_info.promoted_to = ChessPiece.Rook
        elif promotion_char == 'b':
            move_info.promoted_to = ChessPiece.Bishop
        elif promotion_char == 'n':
            move_info.promoted_to = ChessPiece.Knight
        else:
            raise ValueError("Invalid promotion piece in coordinate notation.")
    elif move_info.piece_type is ChessPiece.Pawn and (move_info.target_y == 0 or move_info.target_y == 7):
        move_info.promoted_to = ChessPiece.Queen

def retrieve_additional_move_information(move_info):
    move_info.source_piece = move_info.chess_board[move_info.source_x][move_info.source_y]
    move_info.target_piece = move_info.chess_board[move_info.target_x][move_info.target_y]
    move_info.distance_x = move_info.target_x - move_info.source_x
    move_info.distance_y = move_info.target_y - move_info.source_y
    move_info.direction_x = 1 if move_info.distance_x > 0 else -1 if move_info.distance_x < 0 else 0
    move_info.direction_y = 1 if move_info.distance_y > 0 else -1 if move_info.distance_y < 0 else 0

    if(move_info.distance_x != 0 and move_info.source_piece is not None and move_info.source_piece.type == ChessPiece.Pawn and move_info.target_piece == None):
        if(move_info.en_passant_target == (move_info.target_x, move_info.target_y)):
            move_info.is_enpassant = True
            move_info.chess_board[move_info.target_x][move_info.source_y] = None #!Remove the captured pawn for proper validation!

    #!Castling is to be performed after validation!

